packages/compass-app-stores/src/stores/instance-store.ts

```diff
@@ -0,0 +1,669 @@
+   */
+  async function fetchCollectionMetadata(ns: string) {
+    const { instance, dataService } = store.getState();
+    const { database, collection } = toNS(ns);
+
+    if (!instance || !dataService) {
+      debug(
+        'Trying to fetch collection metadata without the model or dataService in the state'
+      );
+      return;
+    }
+
+    const coll = await instance.getNamespace({
+      database,
+      collection,
+      dataService,
+    });
+
+    return await coll?.fetchMetadata({ dataService });
+  }
+
+  async function refreshNamespace({
+    ns,
+    database,
+  }: {
+    ns: string;
+    database: string;
+  }) {
+    const instance = getInstance();
+    if (!instance.databases.get(database)) {
+      await instance.fetchDatabases({ dataService, force: true });
+    }
+    const db = instance.databases.get(database);
+    if (!db?.collections.get(ns)) {
+      await db?.fetchCollections({ dataService, force: true });
+    }
+    await refreshNamespaceStats(ns);
+  }
+
+  async function refreshNamespaceStats(ns: string) {
+    const instance = getInstance();
+    const { database } = toNS(ns);
+    const db = instance.databases.get(database);
+    const coll = db?.collections.get(ns);
+    // We don't care if this fails
+    await Promise.allSettled([
+      db?.fetch({ dataService, force: true }),
+      coll?.fetch({ dataService, force: true }),
+    ]);
+  }
+
+  const connectionString = dataService.getConnectionString();
+  const firstHost = connectionString.hosts[0] || '';
+  const [hostname, port] = firstHost.split(':');
+
+  const initialInstanceProps: Partial<MongoDBInstanceProps> = {
+    _id: firstHost,
+    hostname: hostname,
+    port: port ? +port : undefined,
+    topologyDescription: getTopologyDescription(
+      dataService.getLastSeenTopology()
+    ),
+  };
+  const instance = new MongoDBInstance(
+    initialInstanceProps as MongoDBInstanceProps
+  );
+  if ((globalThis as any).hadronApp) {
+    (globalThis as any).hadronApp.instance = instance;
+  }
+
+  debug('instance-created');
+  appRegistry.emit('instance-created', { instance });
+
+  store.dispatch(changeDataService(dataService));
+  store.dispatch(changeInstance(instance));
+
+  void refreshInstance({
+    fetchDatabases: true,
+    fetchDbStats: true,
+  });
+
+  function onTopologyDescriptionChanged({
+    newDescription,
+  }: {
+    newDescription: ReturnType<DataService['getLastSeenTopology']>;
+  }) {
+    (instance as any).set({
+      topologyDescription: getTopologyDescription(newDescription),
+    });
+  }
+  dataService.on('topologyDescriptionChanged', onTopologyDescriptionChanged);
+  cleanup.push(() =>
+    dataService.off?.(
+      'topologyDescriptionChanged',
+      onTopologyDescriptionChanged
+    )
+  );
+
+  cleanup.push(() => {
+    const hadronApp = (globalThis as any).hadronApp;
+    if (hadronApp?.instance === instance) {
+      hadronApp.instance?.removeAllListeners();
+      hadronApp.instance = null;
+    }
```

Comment:
I know that this is existing code, but we should probably use the one from the current store in the clean up instead of global one? I'd assume that as soon as all other plugins are using instance service locator, we'd be able to remove this global instance assignment altogether 🤔

Reply:
Right, sorry, I missed that the `removeAllListeners()` was called on the global one rather than the current one. Fixed that + added a TODO to make sure we don’t forget to remove the global property :)